# Practice Case Day 1
MongoDB adalah database dokumen, yang artinya **menyimpan data dalam dokumen mirip JSON**. Berdasarkan beberapa percobaan, mongoDB jauh lebih ekspresif dan kuat daripada model baris / kolom tradisional.

Dalam menjalankan analisis menggunakan NoSQL di tahapan kali ini menggunakan dua cluster dengan dua kondisi :<br>
1. Cluster untuk mengecek **collection movies** yaitu menggunakan mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority<br>
2. Cluster untuk membuat collection baru menggunakan mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority
<br>

### Tahapan dalam Analisis Menggunakan NoSQL
1. Membuat collection baru dengan nama clean_movies yang sama persis dengan movies collection pada database sample_mflix dengan memakai collection movies_intial (projecting_queries pymongo)<br>
2. Validasi collection yang telah dibuat dengan parameter sbb :<br>
        a. Semua document pada clean_movies dan movie sama
        b. Banyak document pada clean_movies dan movie sama
        c. Semua fields pada clean_movies ada pada movie
        d. Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

Sertakan scriptnya dengan format judul script : practice1_nama.ipynb

# GET STARTED

### 1. Install dan Import Package 
Pada tahapan ini beberapa packages yang dibutuhkan adalah package pymongo, pymongo[srv], dan pprint.

In [1]:
!pip install pymongo
!pip install pymongo[srv]
!pip install pprint

You are using pip version 9.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


You are using pip version 9.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


You are using pip version 9.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [2]:
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

import datetime
import pandas as pd
import re

### 2. Connect to the Client

In [3]:
## client untuk collection
client = MongoClient("mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority")
## client untuk mencocokkan antara collection movies_initial dengan movies
clientmovie = MongoClient("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority")

print(client)

MongoClient(host=['cluster0-shard-00-00-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-01-miqju.gcp.mongodb.net:27017', 'cluster0-shard-00-02-miqju.gcp.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='Cluster0-shard-0', ssl=True)


In [4]:
### show all db in cluster
# print(client1.list_database_names())

db=client['sample_mflix'] #database di dbmongo
col=db.movies_initial ## collection cluster baru

dbmovie=clientmovie['sample_mflix'] #database di dbmongo
colmovie=dbmovie.movies ## collection movies

cols=db
colsmovie=dbmovie

print(db.list_collection_names())
print(dbmovie.list_collection_names())

['clean_movies_Putrisqiana', 'clean_movies_tara', 'clean_movies_bunga', 'clean_movies_zumar', 'clean_movies_jul', 'movies_initial', 'clean_movies_putrisqiana', 'clean_movies_firdaus', 'clean_movies_dwilaras', 'clean_movies_nurlailiis', 'clean_movies_faizah']
['clean_movies_dwilaras', 'clean_movies_nurlailiis', 'clean_movies_devita', 'movies_initial', 'clean_movies_frans', 'clean_movies_tara', 'clean_movies_tunggal', 'clean_movies', 'users', 'clean_movies_nurrimah', 'movies_testtopik', 'movies_scratch', 'clean_movies_topik', 'movies', 'clean_movies_afifah', 'clean_movies_andreas', 'clean_movies_indra_1']


In [5]:
### mengambil satu dokumen random
pprint.pprint(col.find_one())

{'_id': ObjectId('5daffb1b986c4adedb754af9'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a '
             'dress with a hoop skirt and white high-heeled pumps, Carmencita '
             'does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress '
         'with a hoop skirt and white high-heeled pumps, Carmencita does a '
         'dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer':

### 3. Sneak Peaks Dataset

In [6]:
### menampilkan semua key yang ada dalam COLLECTION MOVIES_INITIAL ke vertical list
b = [list(doc.keys()) for doc in col.find({})]
b = set([j for i in b for j in i])
b

{'_id',
 'awards',
 'cast',
 'country',
 'director',
 'fullplot',
 'genre',
 'imdbID',
 'imdbRating',
 'imdbVotes',
 'language',
 'lastupdated',
 'metacritic',
 'plot',
 'poster',
 'rating',
 'released',
 'runtime',
 'title',
 'type',
 'writer',
 'year'}

In [8]:
### menampilkan semua key yang ada dalam COLLECTION MOVIES ke vertical list
b = [list(doc.keys()) for doc in colmovie.find({})]
b = set([j for i in b for j in i])
b

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [9]:
### mengecek value key dalam masing-masing dokumen di COLLECTION MOVIES
pd.DataFrame(list(colmovie.find().limit(3)))

_id  \
0  573a1390f29313caabcd5a93   
1  573a1390f29313caabcd60e4   
2  573a1391f29313caabcd71f5   

                                                plot                  genres  \
0  Christ takes on the form of a pacifist count t...                 [Drama]   
1  Charlie is an immigrant who endures a challeng...  [Short, Comedy, Drama]   
2  A young man, unaccustomed to children, must ac...         [Comedy, Short]   

   runtime                                               cast  \
0       78  [Howard C. Hickman, Enid Markey, Lola May, Kat...   
1       30  [Charles Chaplin, Edna Purviance, Eric Campbel...   
2       35     [Harold Lloyd, Mildred Davis, Anna Mae Bilson]   

   num_mflix_comments                                             poster  \
0                 2.0  https://m.media-amazon.com/images/M/MV5BMjAwNT...   
1                 3.0  https://m.media-amazon.com/images/M/MV5BMTNkYW...   
2                 NaN  https://m.media-amazon.com/images/M/MV5BYjgzYz...   

           title                                           fullplot countries  \
0   Civilization  Allegorical film about peace. A king starts a ...     [USA]   
1  The Immigrant  Charlie is on his way to the USA. He wins in a...     [USA]   
2   Now or Never  Mary is looking after a young child whose pare...     [USA]   

   ...                                          directors  \
0  ...  [Reginald Barker, Thomas H. Ince, Raymond B. W...   
1  ...                                  [Charles Chaplin]   
2  ...                      [Fred C. Newmeyer, Hal Roach]   

                                         writers  \
0                          [C. Gardner Sullivan]   
1                                            NaN   
2  [H.M. Walker (titles), Sam Taylor (scenario)]   

                                              awards  \
0    {'wins': 1, 'nominations': 0, 'text': '1 win.'}   
1    {'wins': 1, 'nominations': 0, 'text': '1 win.'}   
2  {'wins': 0, 'nominations': 1, 'text': '1 nomin...   

                     lastupdated  year  \
0  2015-04-30 02:13:27.863000000  1916   
1  2015-09-17 04:52:02.293000000  1917   
2  2015-06-25 00:02:16.613000000  1921   

                                         imdb   type  \
0   {'rating': 6.3, 'votes': 162, 'id': 6517}  movie   
1  {'rating': 7.8, 'votes': 4680, 'id': 8133}  movie   
2  {'rating': 6.8, 'votes': 489, 'id': 12512}  movie   

                                            tomatoes  languages    rated  
0  {'viewer': {'rating': 0.0, 'numReviews': 7}, '...        NaN      NaN  
1  {'viewer': {'rating': 4.1, 'numReviews': 636, ...  [English]  UNRATED  
2  {'viewer': {'rating': 3.8, 'numReviews': 3}, '...  [English]   PASSED  

[3 rows x 21 columns]

In [10]:
### mengecek value key dalam masing-masing dokumen di COLLECTION MOVIES_INITIAL
pd.DataFrame(list(col.find().limit(3)))

_id  imdbID             title  year     rating  \
0  5daffb1b986c4adedb754af9       1        Carmencita  1894  NOT RATED   
1  5daffb1b986c4adedb754afa       5  Blacksmith Scene  1893    UNRATED   
2  5daffb1b986c4adedb754afb       3    Pauvre Pierrot  1892              

  runtime                     genre    released              director writer  \
0   1 min        Documentary, Short              William K.L. Dickson          
1   1 min                     Short  1893-05-09  William K.L. Dickson          
2   4 min  Animation, Comedy, Short  1892-10-28         �mile Reynaud          

   ... imdbRating imdbVotes  \
0  ...        5.9      1032   
1  ...        6.2      1189   
2  ...        6.7       566   

                                              poster  \
0  http://ia.media-imdb.com/images/M/MV5BMjAzNDEw...   
1                                                      
2                                                      

                                                plot  \
0  Performing on what looks like a small wooden s...   
1  Three men hammer on an anvil and pass a bottle...   
2  One night, Arlequin come to see his lover Colo...   

                                            fullplot language country  awards  \
0  Performing on what looks like a small wooden s...              USA           
1  A stationary camera looks at a large anvil wit...              USA  1 win.   
2  One night, Arlequin come to see his lover Colo...           France           

                     lastupdated   type  
0  2015-08-26 00:03:45.040000000  movie  
1  2015-08-26 00:03:50.133000000  movie  
2  2015-08-12 00:06:02.720000000  movie  

[3 rows x 22 columns]

Berdasarkan hasil *sneak peak data* beberapa variabel yang terdapat di *collection*  movies, namun tidak ada di *collection*  movies_intial adalah:
1. num_mflix_comments
2. tomatoes

Dan terdapat variabel di *collection* movies_initial yang tidak terdapat di *collection* movies, yaitu metacritic.

Serta terdapat beberapa perbedaan **nama variabel, tipe data, jumlah data, dan lain sebagainya.** Sehingga sebelum membuat *collection* baru yang mengambil database sample_mflix dan menggunakan *collection* movies_intial, kemudian mencocokkan dengan *collection* movies, perlu merubah beberapa isian data agar sesuai.

### 4. Pipeline
Membuat pipepline baru dalam *collection* dengan menggunakan nama **clean_movies_nurlailiis**.

In [11]:
imdb = list(colmovie.find({}, {'_id':0, 'imdb':1}))
imdb_list=[]
print(len(imdb))
for check in imdb:
    imdb_list.append(list(check.values())[0])

23541


Perbedaan jumlah data di movies terdapat 23541 data, sedangkan di movies_initial terdapat 46014

In [12]:
imdb_list[:5]

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

In [13]:
pipeline = [
    {
        '$project' : {
            'title' : 1,
            'year' : 1,
            'directors' : {'$split' : ['$director', ', ']},
            'cast' : {'$split' : ['$cast', ', ']},
            'writers' : {'$split' : ['$writer', ', ']},
            'genres' : {'$split' : ['$genre', ', ']},
            'languages' : {'$split' : ['$language', ', ']},
            'countries' : {'$split' : ['$country', ', ']},
            'plot' : 1,
            # renaming the keys
            'fullplot' : '$fullplot',
            'rated' : '$rating',
            'released': {
                '$cond': {
                    'if': {'$ne': ["$released", ""]},
                    'then': {
                        '$dateFromString': {
                            'dateString': "$released"
                        }
                    },
                    'else': ""}},
            'runtime': {
                "$cond": 
                            [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]
                       },
            'poster' : 1,
            ### adding embeded documents ###
            ### grouping the several keys into embeded documents by one keys ###
            'imdb' : {
                'id' : '$imdbID',
                'rating' : '$imdbRating',
                'votes' : '$imdbVotes',
                },
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            'awards' : {
                'win' : {'$arrayElemAt' : [{'$split': [{'$arrayElemAt':[{'$split': ['$awards','win']}, -2]}, ' ']}, -1]},
                'nomination' : {'$arrayElemAt' : [{'$split': [{'$arrayElemAt':[{'$split': ['$awards','nomination']}, -2]}, ' ']}, -1]},
                'text' : '$awards',
                },
            'type' : 1,
            'lasUpdated' : '$lastupdated',
            'num_mflix_comments': '', ## tidak ada di collection movies_initial
            'tomatoes': { ## tidak ada di collection movies_initial
                'viewer' : {
                            'rating' : '',
                            'numReviews' : '',
                            'meter' : '',
                            },
                'lasUpdated' : '',
                },
        }
    },{
        ## mencocokakan imdb di movies dan movies_initial
        '$match':{'imdb': 
                      {'$in':[dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}
                 }
    },
    {
        '$out' : 'clean_movies_nurlailiis'
    }
]

clear_output ()
pprint.pprint(list(client.sample_mflix.movies_initial.aggregate(pipeline)))

[]


In [14]:
db=client['sample_mflix'] #database di dbmongo
colclean=db.clean_movies_nurlailiis ## collection cluster baru

Terdapat beberapa variabel di *collection* movies initial atau clean_movies yang memiliki value yang berbeda dengan *collection* movies, diantaranya:
1. runtime
2. awards

### 5. Updating Documents

In [15]:
runtime_pat = re.compile(r'([0-9]+) min')

for movie in client.sample_mflix.clean_movies_nurlailiis.find({}):

    fields_to_set = {}
    fields_to_unset = {}

    for k,v in movie.copy().items():
        if v == "" or v == [""]:
            del movie[k]
            fields_to_unset[k] = ""

    if 'runtime' in movie:
        m = runtime_pat.match(movie['runtime']) 
        if m:
            fields_to_set['runtime'] = int(m.group(1))

    update_doc = {}
    if fields_to_set:
        update_doc['$set'] = fields_to_set
    if fields_to_unset:
        update_doc['$unset'] = fields_to_unset
    pprint.pprint(update_doc)

    db.clean_movies_nurlailiis.update_one({'_id': movie['_id']}, update_doc)


{'$unset': {'languages': '',
            'num_mflix_comments': '',
            'poster': '',
            'writers': ''}}
{'$unset': {'num_mflix_comments': '', 'writers': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': '', 'writers': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'fullplot': '', 'num_mflix_comments': '', 'writers': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'fullplot': '', 'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'languages': '', 'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'languages': '', 'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': ''}}
{'$unset': {'languages': '', 'num_mflix_comments': ''}}
{'$unset': {'num_mfl

{'$unset': {'fullplot': '',
            'num_mflix_comments': '',
            'plot': '',
            'poster': '',
            'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': ''}}
{'$unset': {'fullplot': '', 'num_mflix_comments': '', 'plot': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'languages': '', 'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'languages': '',
            'num_mflix_comme

{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''

{'$unset': {'num_mflix_comments': '', 'rated': '', 'writers': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'poster': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '

{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'fullplot': '',
            'num_mflix_comments': '',
            'plot': '',
            'poster': '',
            'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': ''}}
{'$unset': {'num_mflix_comments': '', 'rated': ''}}
{'$unset': {'

TypeError: expected string or bytes-like object

In [16]:
list(db.clean_movies_nurlailiis.find({'title':'Blacksmith Scene'})) #UNTUK VALIDASI DOCUMENT BARU YANG UDAH DI INSERT KE COLLECTION CLEAN_MOVIES_NURLAILIIS

[{'_id': ObjectId('5daffb1b986c4adedb754afa'),
  'awards': {'nomination': None, 'text': '1 win.', 'win': ''},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lasUpdated': '2015-08-26 00:03:50.133000000',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': '1',
  'title': 'Blacksmith Scene',
  'tomatoes': {'lasUpdated': '',
   'vie

In [17]:
list(dbmovie.movies.find({'title':'Blacksmith Scene'})) #UNTUK VALIDASI DOCUMENT BARU YANG UDAH DI INSERT KE COLLECTION MOVIES

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith behind it and one on either side. The smith in the middle draws a heated metal rod from the fire, places it on the anvil, and all three begin a rhythmic hammering. After several blows, the metal goes back in the fire. One smith pulls out a bottle of beer, and they each take a swig. Then, out comes the glowing metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released': datetime.datetime(1893, 5, 9, 0, 0),
  'runtime': 1,
  'title': 'Blacksmith Scene',
  'tomatoes': {'

# VALIDASI

### a. Semua document pada clean_movies dan movie sama

In [52]:
docmovie = colmovie.find().sort('title')
docclean = colclean.find().sort('title')

#### Pembuktian

In [55]:
if docmovie==docclean:
    print("Semua document pada clean_movies dan movie sama")
else:
    print("Semua document pada clean_movies dan movie tidak sama")

Semua document pada clean_movies dan movie tidak sama


Dokumen antara kedua *collection* **tidak sama** dikarenakan terdapat dokumen yang duplikasi sebanyak tiga kali di *collection* movies, dan akan **dibuktikan pada validasi poin b.**

### b. Banyak document pada clean_movies dan movie sama

In [19]:
imdb = list(dbmovie.movies.find({}, {'_id':0, 'imdb':1}))
print("Banyak dokumen di collection movies sebanyak : ",len(imdb))

Banyak dokumen di collection movies sebanyak :  23541


In [20]:
imdb = list(db.clean_movies_nurlailiis.find({}, {'_id':0, 'imdb':1}))
print("Banyak dokumen di collection clean_movies_nurlailiis sebanyak : ",len(imdb))

Banyak dokumen di collection clean_movies_nurlailiis sebanyak :  23539


#### Pembuktian kenapa jumlahnya berbeda

In [21]:
grup = list(db.clean_movies_nurlailiis.aggregate([{"$sortByCount":"$imdb"}]))
grup2 = list(dbmovie.movies.aggregate([{"$sortByCount":"$imdb"}]))

for i in grup2:
  if i not in grup:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


*Collection* movies memiliki jumlah *document* sebanyak 23541, sedangkan *collection* movies_initial atau *collection* baru clean_movies_nurlailiis hanya memiliki 23539 *document*. Berdasarkan code diatas diketahui bahwa movies dengan **id 5 di *collection* movies duplikasi sebanyak tiga kali** dan **hanya diambil satu document** di *collection* movies_initial atau *collection* baru clean_movies_nurlailiis. 

### c. Semua fields pada clean_movies ada pada movie

In [23]:
### menampilkan semua key yang ada dalam COLLECTION MOVIES ke vertical list 
field_movies = set([j for i in [list(doc.keys()) for doc in colmovie.find({})] for j in i])
field_movies

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [24]:
### menampilkan semua key yang ada dalam COLLECTION CLEAN_MOVIES ke vertical list 
field_clean = set([j for i in [list(doc.keys()) for doc in colclean.find({})] for j in i])
field_clean

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lasUpdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

#### Pembuktian hasilnya sama

In [32]:
### pencocokan
for i in field_movies:
  if i not in field_clean:
    result=0
  else:
    result=1
    
if result==0:
    print("Semua fields pada clean_movies ada pada movie tidak sama")
else:
    print("Semua fields pada clean_movies ada pada movie sama")

Semua fields pada clean_movies ada pada movie sama


Semua *fields* yang terdapat di *collection* movies dan *collection* clean_movies **sama**

### d. Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama
Menggunakan db.collection.find(query, projection) untuk memilih dokumen dalam koleksi atau melihat dan mengembalikan ke dokumen yang dipilih.

In [82]:
valmovie = colmovie.find().sort('title')
valclean = colclean.find().sort('title')

#### Pembuktian

In [83]:
### pencocokan
if valmovie==valclean:
    print("Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama")
else:
    print("Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang tidak sama")

Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang tidak sama


Value antara kedua *collection* **tidak sama** dikarenakan beberapa variabel valuenya beda, diantaranya 